### PR TITLE
Shape inference for dot

### PR DESCRIPTION
**Pull request by farizrahman4u**

### CODE REVIEW

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -307,11 +307,15 @@ def moving_average_update(variable, value, momentum):
...
     if is_sparse(x):
-        return th_sparse_module.basic.structured_dot(x, y)
+        out = th_sparse_module.basic.structured_dot(x, y)
     else:
-        return T.dot(x, y)
+        out = T.dot(x, y)
+    if hasattr(x, '_keras_shape') and hasattr(y, '_keras_shape'):
+        np_x = np.zeros(x._keras_shape)
+        np_y = np.zeros(y._keras_shape)
+        out._keras_shape = np.dot(np_x, np_y).shape
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Should we rely have to rely on `np.dot`? Besides the fact that it's expensive, it's also a black box. The logic should be inferable from reading the code.

**Comment by fchollet** [comment]
Looks good, also please add a test for the shape inference in backend_test.py.

**Comment by farizrahman4u** [comment]
I guess this is already being tested? Do you want me to add more test cases?

**Comment by farizrahman4u** [comment]
See :
https://github.com/fchollet/keras/blob/keras-2/tests/keras/backend/backend_test.py#L81
 https://github.com/fchollet/keras/blob/keras-2/tests/keras/backend/backend_test.py#L53

**Comment by fchollet** [comment]
Right. LGTM

**Pull request merged**